services/pull/merge.go

```diff
@@ -229,33 +240,89 @@ func Merge(pr *models.PullRequest, doer *models.User, baseGitRepo *git.Repositor
 		}
 		// Rebase before merging
 		if err := git.NewCommand("rebase", "-q", baseBranch).RunInDirPipeline(tmpBasePath, nil, &errbuf); err != nil {
+			// Rebase will leave a REBASE_HEAD file in .git if there is a conflict
+			if _, statErr := os.Stat(filepath.Join(tmpBasePath, ".git", "REBASE_HEAD")); statErr == nil {
+				// The original commit SHA1 that is failing will be in .git/rebase-apply/original-commit
+				commitShaBytes, readErr := ioutil.ReadFile(filepath.Join(tmpBasePath, ".git", "rebase-apply", "original-commit"))
+				if readErr != nil {
+					// Abandon this attempt to handle the error
+					return fmt.Errorf("git rebase [%s -> %s]: %s", headRepoPath, tmpBasePath, errbuf.String())
+				}
+				return models.ErrRebaseConflicts{
+					Style:     mergeStyle,
+					CommitSHA: strings.TrimSpace(string(commitShaBytes)),
+					StdOut:    outbuf.String(),
+					StdErr:    errbuf.String(),
+					Err:       err,
+				}
+			}
 			return fmt.Errorf("git rebase [%s -> %s]: %s", headRepoPath, tmpBasePath, errbuf.String())
 		}
+		outbuf.Reset()
 		// Checkout base branch again
 		if err := git.NewCommand("checkout", baseBranch).RunInDirPipeline(tmpBasePath, nil, &errbuf); err != nil {
 			return fmt.Errorf("git checkout: %s", errbuf.String())
 		}
 		// Merge fast forward
-		if err := git.NewCommand("merge", "--ff-only", "-q", stagingBranch).RunInDirPipeline(tmpBasePath, nil, &errbuf); err != nil {
+		if err := git.NewCommand("merge", "--ff-only", "-q", stagingBranch).RunInDirPipeline(tmpBasePath, &outbuf, &errbuf); err != nil {
+			// Merge will leave a MERGE_HEAD file in the .git folder if there is a conflict
+			if _, statErr := os.Stat(filepath.Join(tmpBasePath, ".git", "MERGE_HEAD")); statErr == nil {
+				// We have a merge conflict error
+				return models.ErrMergeConflicts{
+					Style:  mergeStyle,
+					StdOut: outbuf.String(),
+					StdErr: errbuf.String(),
+					Err:    err,
+				}
+			}
 			return fmt.Errorf("git merge --ff-only [%s -> %s]: %s", headRepoPath, tmpBasePath, errbuf.String())
 		}
+		outbuf.Reset()
```

Comment:
So if errbuf also need to be reset?

Reply:
It's been working so far without errbuf resets - AFAIU git will only write on stderr if it returns an error status.

Reply:
Defensive programming? May be a hook will add something in stderr. I do that in commands where I don't want to clutter the normal output (e.g. `find` may throw warnings that are not fatal on stderr, so your list of files in stdout is clean for processing).

Reply:
I've put the resets in.